### PR TITLE
enable aes cfb mode

### PR DESCRIPTION
**Pull request by la-luo**
## Description of changes in release / Impact of release:
OpenPGP is using AES CFB mode and we are missing it in larky

## Documentation
(insert text here)

### Is this a breaking change?
- [ ] Yes
- [x] No

### If you answered Yes then describe why is it so
(insert text here if applicable)

### Is there a way to disable the change?
- [x] Use previous release
- [ ] Use a feature flag
- [ ] No

#### Additional details go here
(insert text here if applicable)

### CODE REVIEW

File: larky/src/main/resources/vendor/Crypto/Cipher/_mode_cfb.star [hunk 1]
```diff
@@ -212,34 +145,71 @@ def CfbMode(block_cipher, iv, segment_size):
...
+        self.IV = self.iv
+        """Alias for `iv`"""
+
+        self._next = [ self.encrypt, self.decrypt ]
+        return self
+
+    self = __init__(self, block_cipher, iv, segment_size)
     return self
 
+def divmod(x, y):
```
Comment (mahmoudimus) [inline larky/src/main/resources/vendor/Crypto/Cipher/_mode_cfb.star:211]: We have a built in divmod as well. This is not necessary.

File: larky/src/test/resources/vendor_tests/pycryptodome/Crypto/Cipher/test_CFB.star [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+load("@stdlib//unittest","unittest")
+load("@vendor//Crypto/Cipher/AES", AES="AES")
+load("@vendor//asserts","asserts")
+
+def b(s):
+    return builtins.bytes(s, encoding="utf-8")
+
+eq = asserts.eq
+
+def CfbTests_test():
```
Comment (mahmoudimus) [inline larky/src/test/resources/vendor_tests/pycryptodome/Crypto/Cipher/test_CFB.star:12]: You will need a lot more comprehensive test suite than this. There are usually many edge cases that we have to run through but I think this is a good first pass.

File: larky/src/main/resources/vendor/Crypto/Cipher/_mode_cfb.star [hunk 3]
```diff
@@ -278,13 +248,18 @@ def _create_cfb_cipher(factory, **kwargs):
...
 
     if kwargs:
-        fail(" TypeError(\"Unknown parameters for CFB: %s\" % str(kwargs))")
-    return CfbMode(cipher_state, iv, segment_size_bytes)
+        fail("TypeError: Unknown parameters for CFB: %s" % str(kwargs))
+
+    return _CfbMode(cipher_state, iv, segment_size_bytes)
+
+CfbMode = larky.struct(
+    _create_cfb_cipher=_create_cfb_cipher,
```
Comment (mahmoudimus) [inline larky/src/main/resources/vendor/Crypto/Cipher/_mode_cfb.star:263]: You need to add a `__name__` here as well

**Review by mahmoudimus — CHANGES_REQUESTED** [review]

File: larky/src/main/resources/vendor/Crypto/Cipher/__init__.star [hunk 4]
```diff
@@ -36,7 +36,7 @@ load("@vendor//Crypto/Cipher/_mode_gcm", GcmMode="GcmMode")
 
 _modes = { 1: EcbMode._create_ecb_cipher,
            2: CbcMode._create_cbc_cipher,
-           # 3:_create_cfb_cipher,
+           3: CfbMode._create_cfb_cipher,
            # 5:_create_ofb_cipher,
            # 6:_create_ctr_cipher,
            # 7:_create_openpgp_cipher,
```
[the remark this answers is not in the document]
Comment (la-luo) [reply larky/src/main/resources/vendor/Crypto/Cipher/__init__.star:42]: yes, otherwise we cannot invoke aes with this new cipher mode

File: larky/src/main/resources/vendor/Crypto/Cipher/_mode_cfb.star [hunk 5]
```diff
@@ -65,57 +50,7 @@ def CfbMode(block_cipher, iv, segment_size):
...
-        self.iv = _copy_bytes(None, None, iv)
-        """The Initialization Vector originally used to create the object.
-        The value does not change."""
-
-        self.IV = self.iv
-        """Alias for `iv`"""
-
-        self._next = [ self.encrypt, self.decrypt ]
-    self = __init__(block_cipher, iv, segment_size)
+    self = larky.mutablestruct(__class__='CfbMode')
```
[the remark this answers is not in the document]
Comment (la-luo) [reply larky/src/main/resources/vendor/Crypto/Cipher/_mode_cfb.star]: done

File: larky/src/main/resources/vendor/Crypto/Cipher/_mode_cfb.star (hunk 1 above)
Comment (la-luo) [reply larky/src/main/resources/vendor/Crypto/Cipher/_mode_cfb.star]: removed

File: larky/src/test/resources/vendor_tests/pycryptodome/Crypto/Cipher/test_CFB.star (hunk 2 above)
Comment (la-luo) [reply larky/src/test/resources/vendor_tests/pycryptodome/Crypto/Cipher/test_CFB.star]: added more tests

File: larky/src/main/resources/vendor/Crypto/Cipher/_mode_cfb.star (hunk 3 above)
Comment (la-luo) [reply larky/src/main/resources/vendor/Crypto/Cipher/_mode_cfb.star]: done

File: larky/src/main/resources/vendor/Crypto/Cipher/__init__.star (hunk 4 above)
[the remark this answers is not in the document]
Comment (mahmoudimus) [reply larky/src/main/resources/vendor/Crypto/Cipher/__init__.star:42]: Sorry, let me clarify. Do you think we need to also implement `_create_openpgp_cipher` as well?

File: larky/src/test/resources/vendor_tests/pycryptodome/Crypto/Cipher/test_CFB.star [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+
+def _testsuite():
+    _suite = unittest.TestSuite()
+    _suite.addTest(unittest.FunctionTestCase(CfbTests_test_verify_ciphertext))
+    _suite.addTest(unittest.FunctionTestCase(CfbTests_test_random_loopback))
+    _suite.addTest(unittest.FunctionTestCase(CfbTests_test_incorrect_iv))
+    _suite.addTest(unittest.FunctionTestCase(CfbTests_test_incorrect_segment_size))
+    _suite.addTest(unittest.FunctionTestCase(CfbTests_test_unknown_param))
+    _suite.addTest(unittest.FunctionTestCase(_testsuite))
```
Comment (mahmoudimus) [inline larky/src/test/resources/vendor_tests/pycryptodome/Crypto/Cipher/test_CFB.star:67]: Is this right?

File: larky/src/test/resources/vendor_tests/pycryptodome/Crypto/Cipher/test_CFB.star [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    asserts.assert_fails(lambda : AES.new(key_128, AES.MODE_CFB,
+                    segment_size=3), ".*?ValueError")
+
+
+def CfbTests_test_unknown_param():
+    asserts.assert_fails(lambda : AES.new(key_128, AES.MODE_CFB,
+                    some_param=0), ".*?TypeError")
+
+
+def _testsuite():
```
Comment (mahmoudimus) [inline larky/src/test/resources/vendor_tests/pycryptodome/Crypto/Cipher/test_CFB.star:60]: FWIW, when I meant to add more tests, I meant you had to port over: https://github.com/Legrandin/pycryptodome/blob/master/lib/Crypto/SelfTest/Cipher/test_CFB.py as I did for the GCM, ECB, etc. It's fine for now if you're comfortable w/ this as a first pass. But I would say these tests are insufficient.

File: larky/src/test/resources/vendor_tests/pycryptodome/Crypto/Cipher/test_CFB.star (hunk 6 above)
Comment (la-luo) [reply larky/src/test/resources/vendor_tests/pycryptodome/Crypto/Cipher/test_CFB.star:67]: yes, they all get passed

Comment (mahmoudimus) [reply larky/src/test/resources/vendor_tests/pycryptodome/Crypto/Cipher/test_CFB.star:67]: I think you made a mistake here, line 67 is not right. I don't think there's a `_testsuite` here that is added to that suite.

File: larky/src/test/resources/vendor_tests/pycryptodome/Crypto/Cipher/test_CFB.star (hunk 7 above)
Comment (la-luo) [reply larky/src/test/resources/vendor_tests/pycryptodome/Crypto/Cipher/test_CFB.star:60]: got it... didn't realize that. Can we merge this pr for now for pgp?  I will add that test later

File: larky/src/test/resources/vendor_tests/pycryptodome/Crypto/Cipher/test_CFB.star (hunk 6 above)
Comment (la-luo) [reply larky/src/test/resources/vendor_tests/pycryptodome/Crypto/Cipher/test_CFB.star:67]: ooops you are right let me remove it...

**Review by mahmoudimus — APPROVED** [review]

File: larky/src/test/resources/vendor_tests/pycryptodome/Crypto/Cipher/test_CFB.star (hunk 7 above)
Comment (mahmoudimus) [reply larky/src/test/resources/vendor_tests/pycryptodome/Crypto/Cipher/test_CFB.star:60]: I approved changes, waiting on your tests to pass. Someone else needs to review so we can merge. @moehajj ?

**Review by moehajj — APPROVED** [review]

File: larky/src/test/resources/vendor_tests/pycryptodome/Crypto/Cipher/test_CFB.star (hunk 7 above)
Comment (moehajj) [reply larky/src/test/resources/vendor_tests/pycryptodome/Crypto/Cipher/test_CFB.star:60]: 1 review should be good, just resolve these conversations once done because they will block the merge

File: larky/src/main/resources/vendor/Crypto/Cipher/__init__.star (hunk 4 above)
[the remark this answers is not in the document]
Comment (la-luo) [reply larky/src/main/resources/vendor/Crypto/Cipher/__init__.star:42]: check pycrypto it is standard CFB not openpgp https://github.com/pycrypto/pycrypto/blob/7acba5f3a6ff10f1424c309d0d34d2b713233019/lib/Crypto/Cipher/blockalgo.py#L82
So far it looks like not necessary
